### DISCUSSION TITLE
Help with circular schemas

### DISCUSSION THREAD

**Discussion by wadjoh** [open-circle/valibot#673] [Q&A] [answered]
Hi there! I'm looking into using Valibot for objects that have internal circular relations, but I'm running into issues and hoping I could get some guidance.

For example:
An object will have a User type object, an Auth type object, and a relation object that links them together. The User and Auth objects will hold a reference to the relation object.

```ts
const user = {
  firstName: 'joe',
  $relations: {
    // will be assigned farther down
    UserToAuth: undefined
  }
}

const auth = {
  email: 'joe@example.com',
  password: '1234',
  $relations: {
    // will be assigned farther down
    UserToAuth: undefined
  }
}

const userToAuthRelation = {
  from: user,
  to: auth
}

user.$relations.UserToAuth = userToAuthRelation
auth.$relations.UserToAuth = userToAuthRelation
```

I tried implementing Valibot schemas to validate these objects using `lazy`:

```ts
const UserSchema = v.object({
  firstName: v.string(),
  $relations: v.object({
    UserToCredentials: v.lazy(() => UserToAuthRelationSchema)
  })
})

const AuthSchema = v.object({
  email: v.pipe(v.string(), v.email()),
  password: v.pipe(v.string(), v.minLength(8)),
  $relations: v.object({
    UserToCredentials: v.lazy(() => UserToAuthRelationSchema)
  })
})

type TUserToAuthRelationSchema = {
  from: v.InferOutput<typeof UserSchema>
  to: v.InferOutput<typeof AuthSchema>
}
const UserToAuthRelationSchema: v.GenericSchema<
  TUserToAuthRelationSchema
> = v.pipe(
  v.object({
    from: v.lazy(() => UserSchema),
    to: v.lazy(() => AuthSchema),
  }),
  v.check((input) => !!input && typeof input === 'object')
)
```

I was able to get the TS types working as expected, but when I call `safeParse` with this schema and an input object, I get `Error: Maximum call stack size exceeded`.

I'm wondering if there's something I am doing wrong or if Valibot is not the right solution for this?

I've created a repro example in a few places for convenience:
- [Valibot Playground](https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6B6JuAZQGMALAUxGTgATHtmAA7YDGAQxqAFz0OM1PACqqHlE69+cALyYAdBAoArHhxgAKAN704OYFBUA5ZCB5yjKqOIDmVgCUADT2cAAkUDykyFLKXhjGZhbWdg4O6poAKhAAwlHCYlJk8kYxAF4AnlZB+gB8cJlQOQCCAK4wXABK0bHSYtp8yIFhAL4j44xKsvDtnYO6Bokm5pa2YUPApAmGYMBgPFaJPv5BwUabpEEhYWBoqADu0II7ewdHhidiASFGIOIAGR4306VgAHIEbg5Ir04rIdisUut0o0NM08gVgcVSKVEhVqrU9A0mq0Ot1Yf0FsMxhMRtNUBBSDxDKQIAFCABaDkobAwTRwVDcIZCETiST9VCEEb0GCVA5wLIkiBzckxOFU-RwNI4PAgHYASTE2E0AHkOmAOgAeWUHCDYVGaKl1MIwCAGo2m81Wm08O1wFVO+ijRTKNRo0mdHpqylC-g7ADiwM0wA4VMtYUV4eVZKjfRkVPoDSWu32hzCy2Sa21Dlw+B2+JqgXqDq0seGoRRrvryCqjebAbbULg4w7cBYKBgfPA8FijgAHv4BTBkBwANZwCAYTTYVkPOCaPBQcuGIVrmriC0wJtEuAAQlvF46cAAZM+4D6-Y-4Hof0REZYpXoaVxyyHgVDgL8FHpeA2jRQ1Ly8HolCgQRLS+PxzmQMRKiLLUwlEZwYDcDwvEIUwIB4QhQmDaCUDJeCOkQixnjQmBfG+TDsNw7VLlI8ieAAAR4Od3DAJkT3wKjbnuJ4UNIgBGAAmABmAAWABWAA2AB2MEpODWDNAYmBDBhaNlE1GwDLgsRL1MqJzNkQwlXyHhCmxdADG1Ws9TgQyoGM0cuzozpjKDehkHo2yOnsikLK84NItC6KTLMvMnJczEimAEpLPw3UvH8wKXTdEKuDCmjlEZZlWXZLkeT5KB3zA+AvxFUQJDhSVgNYUCVH8KDQzgKJUDaUhv28ZBjQABWQZxDgHHRkEwqLL0CBh6Wqlk2SsTluSmxqBSmnhZvmwD6GAe0rBGsaTNGjgODA1Am21WiCNcdweE1G7xuML1Uoc9LUGcrNXPcnKcUMHzYsc4HMrcrEIeB6H3qIz6wk28TaqsVHiJ4dag33HEvteqqsZ2wgGQ8TpFwePBvkA4MgA)
- [Repo](https://github.com/wadjoh/valibot-recursion-testing)
- [Stackblitz](https://stackblitz.com/~/github.com/wadjoh/valibot-recursion-testing) (linked to repo above)

**Comment by fabian-hiller** [comment, marked as answer]
I think the main problem is that your data contains a circular reference, which leads to a parse execution that never ends. `lazy` is intended for recursive schemas like binary trees that end at some point. I think we could technically solve this problem by memorizing what `lazy` does, but the main question for me is whether this is outside the scope of Valibot. Maybe you should validate the data before it is mapped to be circular.

**Reply by wadjoh** [reply]
Oh, memoization didn't actually cross my mind. I'll give that a try!

Also, is there a way to stop Valibot from recursing into the objects? I tried using `v.check` on all those schemas thinking it wouldn't try traversing the objects, but I was still getting the same error.

> the main question for me is whether this is outside the scope of Valibot

Yup, that was my thought as well. No problem if it's considered outside of Valibot's scope -- that'd make sense to me.

**Reply by fabian-hiller** [reply]
No, there is currently no way to stop this. You can try to write your own `lazy` schema that remembers and breaks circular references.

Here is our code: https://github.com/fabian-hiller/valibot/blob/main/library/src/schemas/lazy/lazy.ts
